Framework/Algorithms/src/Stitch1D.cpp

```diff
@@ -1,45 +1,95 @@
-#include "MantidAPI/WorkspaceFactory.h"
+#include "MantidAPI/WorkspaceProperty.h"
+#include "MantidHistogramData/HistogramE.h"
+#include "MantidHistogramData/HistogramDx.h"
 #include "MantidHistogramData/HistogramX.h"
 #include "MantidHistogramData/HistogramY.h"
-#include "MantidHistogramData/HistogramE.h"
 #include "MantidKernel/ArrayProperty.h"
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/MultiThreaded.h"
 #include "MantidKernel/PropertyWithValue.h"
 #include "MantidKernel/RebinParamsValidator.h"
-#include "MantidKernel/MultiThreaded.h"
-#include "MantidKernel/BoundedValidator.h"
 
+#include <algorithm>
 #include <boost/tuple/tuple.hpp>
 #include <boost/format.hpp>
-#include <boost/algorithm/string.hpp>
 #include <boost/math/special_functions.hpp>
-#include <algorithm>
+#include <map>
 
-using namespace Mantid::Kernel;
 using namespace Mantid::API;
-using Mantid::HistogramData::HistogramY;
+using namespace Mantid::Kernel;
 using Mantid::HistogramData::HistogramE;
+using Mantid::HistogramData::HistogramDx;
+using Mantid::HistogramData::HistogramX;
+using Mantid::HistogramData::HistogramY;
+using Mantid::HistogramData::Points;
 
 namespace {
-
+/// Returns a tuple holding the first and last x value of the first spectrum and
+/// the lhs and rhs workspace, respectively
 using MinMaxTuple = boost::tuple<double, double>;
-MinMaxTuple calculateXIntersection(MatrixWorkspace_sptr lhsWS,
-                                   MatrixWorkspace_sptr rhsWS) {
+MinMaxTuple calculateXIntersection(MatrixWorkspace_const_sptr &lhsWS,
+                                   MatrixWorkspace_const_sptr &rhsWS) {
   return MinMaxTuple(rhsWS->x(0).front(), lhsWS->x(0).back());
 }
 
+/// Check if a double is not zero and returns a bool indicating success
 bool isNonzero(double i) { return (0 != i); }
+
+/** Sort x axis
+ @param ws :: Input workspace
+ @return A shared pointer to the resulting MatrixWorkspace
+ */
+void sortXAxis(MatrixWorkspace_sptr &ws) {
+  // using a std::multimap (keys are sorted)
+  std::multimap<double, double> sorter;
+  double x_value;
+  const int nHists = static_cast<int>(ws->getNumberHistograms());
+  for (int i = 0; i < nHists; ++i) {
+    for (size_t k = 0; k < ws->size(); ++k) {
+      x_value = ws->x(i)[k];
+      sorter.insert(std::pair<double, double>(x_value, ws->y(i)[k]));
+      sorter.insert(std::pair<double, double>(x_value, ws->e(i)[k]));
+      if (ws->hasDx(i) && !ws->isHistogramData())
+        sorter.insert(std::pair<double, double>(x_value, ws->dx(i)[k]));
+    }
+    size_t ws_size = ws->size();
+    Mantid::MantidVec vecx(ws_size);
+    Mantid::MantidVec vecy(ws_size);
+    Mantid::MantidVec vece(ws_size);
+    Mantid::MantidVec vecdx(ws_size);
+    int l = 0;
+    auto it = sorter.cbegin();
+    while (it != sorter.cend()) {
+      vecx[l] = it->first;
+      vecy[l] = it->second;
+      vece[l] = (++it)->second;
+      if (ws->hasDx(i) && !ws->isHistogramData())
```

Comment:
Why are you not handling `Dx` for histogram data?

Reply:
For stitching histogram data, the weighted mean will be computed for the overlap region. This changes the number of bins and bin edges in this overlap region and then it is not easy to introduce new `Dx` values. My feeling here is that we need to think about this more deeply. At ILL, point data are used for reflectometry and thus, our scientists may not demand further `Dx` handling.
Histogram data having `Dx` can be converted to point data and then handled, if desired.

Reply:
Ok makes sense. Could you put a small comment explaining this in the code, for future reference?

Reply:
Yes, no problem